r/src/tpch-gen.cpp

```diff
@@ -0,0 +1,167 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "./arrow_types.h"
+
+#if defined(ARROW_R_WITH_ARROW)
+
+#include <arrow/compute/api.h>
+#include <arrow/compute/exec/exec_plan.h>
+#include <arrow/compute/exec/tpch_node.h>
+#if defined(ARROW_R_WITH_DATASET)
+#include <arrow/dataset/api.h>
+#include <arrow/dataset/plan.h>
+#include <arrow/filesystem/filesystem.h>
+#include <arrow/filesystem/localfs.h>
+
+namespace compute = ::arrow::compute;
+namespace ds = ::arrow::dataset;
+namespace fs = ::arrow::fs;
+
+static std::shared_ptr<compute::ExecNode> MakeExecNodeOrStop(
+    const std::string& factory_name, compute::ExecPlan* plan,
+    std::vector<compute::ExecNode*> inputs, const compute::ExecNodeOptions& options) {
+  return std::shared_ptr<compute::ExecNode>(
+      ValueOrStop(compute::MakeExecNode(factory_name, plan, std::move(inputs), options)),
+      [](...) {
+        // empty destructor: ExecNode lifetime is managed by an ExecPlan
+      });
+}
+
+// [[arrow::export]]
+std::shared_ptr<arrow::RecordBatchReader> Tpch_Dbgen(
+    const std::shared_ptr<compute::ExecPlan>& plan, double scale_factor,
+    std::string table_name) {
+  auto gen =
+      ValueOrStop(arrow::compute::internal::TpchGen::Make(plan.get(), scale_factor, 1000000));
```

Comment:
I bumped the batch size from [the default of 4096](https://github.com/apache/arrow/blob/bda70875d2957bf1c890ef0457ce7836dd797aa6/cpp/src/arrow/compute/exec/tpch_node.h#L46) to 1M here. IIRC that's ~right for most cases, but would love a second set of eyes to make sure. (Also should that default in tpch_node be 4096 or something higher like this?)

Reply:
Agreed.  For the moment we probably shouldn't go lower than 32K if we are feeding actual TPC-H queries.  1M seems like it would be fine.

Long term I don't think R should have to specify the batch size here so I do lean slightly towards changing the default vs. changing this here.  That way if we want to drop to 4k in the future we can.

CC @save-buffer for second opinion.

Reply:
I seem to remember someone was benchmarking different batch sizes - what was the outcome of that?

But overall I'm indifferent to the default batch size, since we'll have to figure out how the generator will work later anyway once we switch to the morsel/batch model.